### DISCUSSION TITLE
Programmatic way to extract citation key/citekey?

### DISCUSSION THREAD

**Discussion by ReaderGuy42** [retorquere/zotero-better-bibtex#2978] [Q&A]
I'm working on a script that copies quotations from Zotero to Obsidian, where each source has a note called `citekey.md`, e.g. `Chiang2004What.md`

For this I need to first get the citekey of the document I'm working on. So far I've been using `xdotool key ctrl+shift+c` to get the citekey, but this doesn't work reliably, especially for longer quotes.

Is there another way to get the citekey from within a script? I'm using bash on a Linux PC, and very open to suggestions.

This is how I'm currently getting the citekey:
```
echo -n | xclip -selection clipboard
sleep 0.5
xdotool key ctrl+shift+c
sleep 0.2

citekey=$(xclip -o -selection clipboard)
```

The first line is supposed to clear the clipboard because I was having frequent problems with the program instead using previously copied things as the citekey and then of course not finding a note with that name. But now it sometimes tries finding `.md` so I've increased the `sleep` but I'm not sure if that's completely fixed the problem.

**Comment by retorquere** [comment]
And by this you want the citation key of the selected item in Zotero? I could add behaviour to this JSON-RPC endpoint:

https://retorque.re/zotero-better-bibtex/exporting/json-rpc/index.html#itemcitationkeyitem_keys

so that you could call it wilth `item.citationkey('selected')`

**Reply by ReaderGuy42** [reply]
The way I've been using the script, it would be ideal if it could get the citation key of the currently open PDF in Zotero. Is that possible?

I had a look at the JSON-RPC thing earlier, but I don't understand how to set it up. Would you be able to point me towards a tutorial of some kind? 

Thanks!!

**Reply by retorquere** [reply]
there's no setup -- you would just post to an URL, like eg

```
citekey=$(curl http://localhost:23119/better-bibtex/json-rpc -X POST -H "Content-Type: application/json" -H "Accept: application/json" --data-binary '{"jsonrpc": "2.0", "method": "item.citationkey", "params": {"citekeys": "selected" }}')
```

**Reply by ReaderGuy42** [reply]
OK, thanks for the help! But if I put that `curl` in a terminal I get this:
`{"jsonrpc":"2.0","error":{"code":-32602,"message":"item.citationkey: unexpected argument citekeys"},"id":null}`

and if I put the `http...` into a URL bar it says `No endpoint found`

Do I need to enable JSON-RPC in the Zotero or BBT settings somewhere? I wasn't able to find anything there though.

**Reply by retorquere** [reply]
Sorry, that should have been

```
curl http://localhost:23119/better-bibtex/json-rpc -X POST -H "Content-Type: application/json" -H "Accept: application/json" --data-binary '{"jsonrpc": "2.0", "method": "item.citationkey", "params": [ "selected" ]}'
```

**Reply by retorquere** [reply]
> and if I put the http... into a URL bar it says No endpoint found

These links don't support GET, just POST (and also they don't work from a browser)

**Reply by ReaderGuy42** [reply]
So sorry, I'm still getting an error:
```
{"jsonrpc":"2.0","error":{"code":-32602,"message":"/item_keys: type must be array"},"id":null}
```
I tried selecting a number of Zotero items and also opening a PDF but always got that error message. Any idea what it means?

Thanks so much for your help!

**Reply by retorquere** [reply]
I haven't implemented "selected". If you can open an issue I can get you a test build to see if it does what you need.

**Reply by ReaderGuy42** [reply]
Done, thanks!! https://github.com/retorquere/zotero-better-bibtex/issues/2983

**Reply by ReaderGuy42** [reply]
Thanks for the test build! I added a question :)
